compiler/rustc_monomorphize/src/collector.rs

```diff
@@ -344,17 +346,32 @@ pub fn collect_crate_mono_items(
         let inlining_map: MTRef<'_, _> = &mut inlining_map;
 
         tcx.sess.time("monomorphization_collector_graph_walk", || {
-            par_for_each_in(roots, |root| {
-                let mut recursion_depths = DefIdMap::default();
-                collect_items_rec(
-                    tcx,
-                    dummy_spanned(root),
-                    visited,
-                    &mut recursion_depths,
-                    recursion_limit,
-                    inlining_map,
-                );
-            });
+            if is_parallel() {
+                let tcx = DynSend::new(tcx);
+                par_for_each_in(roots, |root| {
+                    let mut recursion_depths = DefIdMap::default();
+                    collect_items_rec(
+                        *tcx,
+                        dummy_spanned(root),
+                        visited,
+                        &mut recursion_depths,
+                        recursion_limit,
+                        inlining_map,
+                    );
+                });
+            } else {
+                for_each_in(roots, |root| {
+                    let mut recursion_depths = DefIdMap::default();
+                    collect_items_rec(
+                        tcx,
+                        dummy_spanned(root),
+                        visited,
+                        &mut recursion_depths,
+                        recursion_limit,
+                        inlining_map,
+                    );
+                });
+            }
```

Comment:
You say that the closure is `FnMut` and cannot be trivially parallelized.
I'm a bit concerned actually: does that mean there is a non-determinacy here, based on the parallel execution order?

Reply:
This is indeed non-deterministic, but last time I looked at the code here everything ended up sorted somewhere.

Reply:
[merge_codegen_units](https://github.com/rust-lang/rust/blob/master/compiler/rustc_monomorphize/src/partitioning/merging.rs#L27)sorts the codegen_units before merging